### PR TITLE
Add TestAspect.checks to allow running aspects within property-based specs

### LINKED ISSUE #9046: Allow running some effect before / after a single sample of check
# What
This is similar to what was asked for in https://github.com/zio/zio/issues/5285 and https://github.com/zio/zio/issues/4007.

Goal would be to have some mechanism to run an effect before/after a single sample of check. This is different to the existing TestAspect.{before,after,around} as these run effects around the _entire test_ which includes the entire test block.

# Why

Sometimes it is useful to test side-effecting programs using property based testing. An example would be specs for a database, where you want to generate the data to insert into the database and assert against later. This can be useful to detect issues that otherwise. One exampe for this is the mismatch between precision for timestamps in postgres (up to micriseconds) and java.time.Instant (up to nanoseconds). Issues such as this are much easier to find using property-based-testing.

An alternative that can be used is to have a custom function that sets up the database, and is also what I'm currently using, something like:
```scala
test("insert and get user") {
  check(GenUser) { user => 
    withPreparedDb {
      for {
        userRepository <- ZIO.service[UserRepository]
        _ <- userRepository.insert(user)
        retrieved <- userRepository.get(user.id)
      } yield assert(retrieved)(equalTo(user))
    }
  }
}
```
The problem with this is that it plays very badly with reusing the tests / running them against multiple targets. For example, if you have a fake implementation of the UserRepository that you want to use for other tests, it is very nice to be able to run the same tests against that one as well. This works great with the existing aspects / provide, but doesn't work at all with setup functions like above without additional machinery.

# How

I think that best UX for this would be additonal aspects that only run code before / after / around a single sample. Somthing like TestAspect.beforeCheck. I.e. the above would become:
```scala
test("insert and get user") {
  check(GenUser) { user => 
    for {
      userRepository <- ZIO.service[UserRepository]
      _ <- userRepository.insert(user)
      retrieved <- userRepository.get(user.id)
    } yield assert(retrieved)(equalTo(user))
  }
} @@ TestAspect.aroundCheck(preparedDb)
```

This could be implemented using a fiber ref. We are already passing data into tests using fiber refs into the tests with the TestConfig, so it also would not be a big departure from what we already have.
We could add a new FiberRef or add a new field to TestConfig, using it to pass an `UIO[Unit]` or `Option[URIO[Scope, Unit]]` into the test. Inside of the checkStream function that actually runs the property based test, this fiberRef would be retrieved and the embedded effect applied to the test code.

---

If you are happy with the above proposal I'm happy to work on this, so just give me a short 👍 / 👎 to let me know.

### PR DESCRIPTION
**Pull request by mschuwalow**
resolves #9046 
/claim #9046

### CODE REVIEW

File: test/shared/src/main/scala/zio/test/TestConfig.scala [hunk 1]
```diff
@@ -48,6 +49,14 @@ trait TestConfig extends Serializable {
...
    */
   def shrinks: Int
+
+  /**
+   * Aspect that should be applied to each check sample.
+   *
+   * NOTE: default implementation for backward compatibility. Remove in next
+   * major version.
+   */
+  def checkSampleAspect: TestAspect.CheckSampleAspect = ZIOAspect.identity
```
Comment (mschuwalow) [inline test/shared/src/main/scala/zio/test/TestConfig.scala]: This guy is a bit unfortunate as it's possible to forget to forward the aspect when creating proxies like in TestAspect.scala, but it's the most backward-compatible approach I could come up with.
Suggestions welcome.

File: test/shared/src/main/scala/zio/test/TestConfig.scala [hunk 2]
```diff
@@ -56,23 +65,57 @@ object TestConfig {
 
   final case class Test(repeats: Int, retries: Int, samples: Int, shrinks: Int) extends TestConfig
 
+  // NOTE: Additonal class for backward compatibility. Rename to `Test` and remove the old one in next major version.
+  final case class Test1(
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestConfig.scala]: Next major version won't likely be for some time. Perhaps rename this to `TestV2` add a `@deprecated("use TestV2", "2.1.8")` annotation to `Test`?

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 3]
```diff
@@ -279,6 +285,62 @@ object TestAspect extends TimeoutVariants {
...
+  def checksZIO[R, E](
+    makeAspect: ZIO[R, E, CheckAspect]
+  ): TestAspect[Nothing, R, E, Any] =
+    new TestAspect[Nothing, R, E, Any] {
+      def some[R1 <: R, E1 >: E](spec: Spec[R1, E1])(implicit trace: Trace): Spec[R1, E1] =
+        spec.transform[R1, E1] {
+          case Spec.TestCase(oldTest, annotations) =>
+            val newTest = makeAspect.mapError(TestFailure.fail).flatMap { aspect =>
+              testConfigWith { oldConfig =>
+                val newConfig = new TestConfig {
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: I think we can use the `TestV2` case class here instead. I don't really know why the `Test` case class wasn't used here but I can't think of a good reason

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 4]
```diff
@@ -737,10 +799,11 @@ object TestAspect extends TimeoutVariants {
       )(implicit trace: Trace): ZIO[R, TestFailure[E], TestSuccess] =
         testConfigWith { old =>
           val testConfig = new TestConfig {
-            val repeats = n
-            val retries = old.retries
-            val samples = old.samples
-            val shrinks = old.shrinks
+            val repeats              = n
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: As above

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 5]
```diff
@@ -807,10 +870,11 @@ object TestAspect extends TimeoutVariants {
       )(implicit trace: Trace): ZIO[R, TestFailure[E], TestSuccess] =
         testConfigWith { old =>
           val testConfig = new TestConfig {
-            val repeats = old.repeats
-            val retries = n
-            val samples = old.samples
-            val shrinks = old.shrinks
+            val repeats              = old.repeats
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: As above

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 6]
```diff
@@ -845,10 +909,11 @@ object TestAspect extends TimeoutVariants {
       )(implicit trace: Trace): ZIO[R, TestFailure[E], TestSuccess] =
         testConfigWith { old =>
           val testConfig = new TestConfig {
-            val repeats = old.repeats
-            val retries = old.retries
-            val samples = n
-            val shrinks = old.shrinks
+            val repeats              = old.repeats
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: As above

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 7]
```diff
@@ -934,10 +999,11 @@ object TestAspect extends TimeoutVariants {
       )(implicit trace: Trace): ZIO[R, TestFailure[E], TestSuccess] =
         testConfigWith { old =>
           val testConfig = new TestConfig {
-            val repeats = old.repeats
-            val retries = old.retries
-            val samples = old.samples
-            val shrinks = n
+            val repeats              = old.repeats
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: As above

**Review by kyri-petrou — COMMENTED** [review]
Thanks, looks great! Just some minor comments below

File: test/shared/src/main/scala/zio/test/TestConfig.scala (hunk 1 above)
Comment (kyri-petrou) [reply test/shared/src/main/scala/zio/test/TestConfig.scala]: Had this issue myself a few times.

Scala needs an annotation on methods that should always be overwritten by concrete implementations 🥲

File: test/shared/src/main/scala/zio/test/TestAspect.scala [hunk 8]
```diff
@@ -21,6 +21,10 @@ import zio.stacktracer.TracingImplicits.disableAutoTrace
 
 import java.util.concurrent.atomic.AtomicReference
 import scala.collection.immutable.SortedSet
+import zio.test.TestAspectPoly
+import zio.System.env
+import zio.test.TestAspectAtLeastR
+import zio.test.TestAspectAtLeastR
```
Comment (kyri-petrou) [inline test/shared/src/main/scala/zio/test/TestAspect.scala]: Leftover imports

**Review by kyri-petrou — COMMENTED** [review]
Sorry for the nitpick 😅
It's good to go after this

**Review by kyri-petrou — APPROVED** [review]
LGTM, thanks!

**Pull request merged**